mindmeld/auto_annotator.py

```diff
@@ -0,0 +1,694 @@
+    """ Annotator class that uses spacy to generate annotations.
+    """
+
+    def __init__(self, app_path, config=None):
+        super().__init__(app_path=app_path, config=config)
+
+        self.model = self.config.get("spacy_model", EN_CORE_WEB_LG)
+        self.nlp = SpacyAnnotator._load_model(self.model)
+        self.duckling = DucklingRecognizer.get_instance()
+        self.ANNOTATOR_TO_DUCKLING_ENTITY_MAPPINGS = {
+            "money": "sys_amount-of-money",
+            "cardinal": "sys_number",
+            "ordinal": "sys_ordinal",
+            "person": "sys_person",
+            "percent": "sys_percent",
+            "distance": "sys_distance",
+            "quantity": "sys_weight",
+        }
+
+    @staticmethod
+    def _load_model(model):
+        """ Load Spacy English model. Download if needed.
+
+        Args:
+            model (str): Spacy model ("en_core_web_sm", "en_core_web_md", or
+                "en_core_web_lg").
+
+        Returns:
+            nlp (spacy.lang.en.English): Spacy language model.
+        """
+
+        if model in [EN_CORE_WEB_SM, EN_CORE_WEB_MD, EN_CORE_WEB_LG]:
+            logger.info("Loading Spacy model %s.", model)
+            try:
+                return spacy.load(model)
+            except OSError:
+
+                logger.warning("%s not found. Downloading the model.", model)
+
+                os.system("python -m spacy download " + model)
+                language_module = importlib.import_module(model)
+                return language_module.load()
+        else:
+            error_msg = (
+                "Unknown Spacy model name: {!r}. Model must be {},"
+                " {}, or {}".format(
+                    model, EN_CORE_WEB_SM, EN_CORE_WEB_MD, EN_CORE_WEB_LG
+                )
+            )
+            raise ValueError(error_msg)
+
+    def valid_entity_check(self, entity):
+        entity = entity.lower().strip()
+        return entity in SPACY_ANNOTATOR_SUPPORTED_ENTITIES
+
+    def parse(self, sentence, entity_types=None):
+        """ Extracts entities from a sentence. Detected entities should are
+        represented as dictionaries with the following keys: "body", "start"
+        (start index), "end" (end index), "value", "dim" (entity type).
+
+        Args:
+            sentence (str): Sentence to detect entities.
+            entity_types (list): List of entity types to parse. If None, all
+                possible entity types will be parsed.
+
+        Returns:
+            entities (list): List of entity dictionaries.
+        """
+        doc = self.nlp(sentence)
+        spacy_entities = [
+            {
+                "body": ent.text,
+                "start": ent.start_char,
+                "end": ent.end_char,
+                "value": {"value": ent.text},
+                "dim": ent.label_.lower(),
+            }
+            for ent in doc.ents
+        ]
+
+        entities = []
+        for entity in spacy_entities:
```

Comment:
Instead of having this huge `elif` block, why cant we have a map of entity type and function. There might be some duplicates, but I think its better for readability.

Reply:
Updated